pyfr/plugins/acoustics.py

```diff
@@ -0,0 +1,822 @@
+            mpn = np.empty((nfaces_, nfacefpts))
+            npn = np.empty((nfaces_, nfacefpts, ndims))
+            fplocs = np.empty((nfaces_, nfacefpts, ndims))
+
+            for ie, eidx in enumerate(eidlist):
+                # save flux pts coordinates:
+                fplocs[ie] = eles.plocfpts[facefpts, eidx]
+                # Unit physical normals and their magnitudes (including |J|)
+                fpts_idx = eles.basis.facefpts[fidx]
+                mpn[ie] = mag_pnorm_fpts[fpts_idx, eidx]
+                npn[ie] = pnorm_fpts[fpts_idx, eidx]
+
+            #area differential
+            qdA = qwts*mpn
+            dA = np.hstack((dA, qdA.reshape(-1)))
+
+            fpts_plocs  = np.vstack((fpts_plocs, fplocs.reshape(-1, ndims)))
+            norm_pnorms = np.vstack((norm_pnorms, npn.reshape(-1, ndims)))
+
+        return [fpts_plocs, norm_pnorms, dA], proj
+
+    def _update_slice(self, intg, m0_dict, eidxs, usoln):
+        # Solution matrices indexed by element type
+        solns = dict(zip(intg.system.ele_types, intg.soln))
+        nvars = self.nvars
+
+        fIo = 0
+        for (etype, fidx), m0 in m0_dict.items():
+            # Get the interpolation operator
+            nfpts, nupts = m0.shape
+
+            # Extract the relevant elements from the solution
+            uupts = solns[etype][...,eidxs[etype, fidx]]
+
+            # Interpolate to the face
+            ufpts = m0 @ uupts.reshape(nupts, -1)
+            ufpts = ufpts.reshape(nfpts, nvars, -1)
+            ufpts = ufpts.swapaxes(0, 1) # nvars, nfpts, nfaces
+            nfaces_pertype = ufpts.shape[-1]
+
+            # get primitive vars
+            pri_ufpts = self.elementscls.con_to_pri(ufpts, self.cfg)
+            fIsize = nfpts*nfaces_pertype
+            fImax = fIo + fIsize
+            for i in range(self.nvars):
+                usoln[i, fIo :fImax] = pri_ufpts[i].reshape(-1)
+
+            fIo = fImax
+
+    def _apply_wall_bc(self, bndusoln, bctype='noslip', nvec=None):
+        gm1 = self.constvars['gamma'] - 1
+        rho = bndusoln[0]
+        u = bndusoln[1 :self.ndims + 1]
+        p = bndusoln[-1]
+
+        #new soln after applying the bc
+        p += 0.5*gm1*rho*np.sum([ui*ui for ui in u], axis=0)
+        if bctype == 'noslip':
+            u *= u*0
+        elif bctype == 'slip':
+            u -= np.sum(u*nvec, axis=0)*nvec
+            p -= 0.5*gm1*rho*np.sum([ui*ui for ui in u], axis=0)
+
+    def _write_data(self, outf, pfft):
+        nwindows = self.fwhsolver.nwindows
+        naver = self.fwhsolver.naver
+        freq = self.fwhsolver.freq
+        nfreq = len(freq)
+
+        wa_arr = np.tile([[str(nwindows)], [str(naver)]], nfreq)
+        for opts, p in zip(self.fwhsolver.obsv_pts, pfft):
+            pmag, pang = np.abs(p), np.angle(p)
+            spdata = np.array([freq, pmag, pang])
+            wdata = wa_arr
+            obploc = np.array([opts.tolist()] * nfreq)
+
+            for op in [*obploc.T]:
```

Comment:
This will be slow

Reply:
Ok, I'll have a look at refactoring this.